Add `include` to CLI options

rpominov opened the pull request:
See #207

rpominov commented on bin/nyc.js:
```diff
@@ -46,6 +46,11 @@ var yargs = require('yargs/yargs')(process.argv.slice(2))
     describe: 'whether or not to instrument all files of the project (not just the ones touched by your test suite)'
   })
+  .option('include', {
+    alias: 'in',
```
`i` is taken unfortunately.

rpominov commented on bin/nyc.js:
```diff
@@ -46,6 +46,11 @@ var yargs = require('yargs/yargs')(process.argv.slice(2))
+  .option('include', {
+    alias: 'in',
+    default: [],
+    describe: 'a list of files that should be covered'
```
Not sure whether this description is good enough.

novemberborn replied:
[From the readme](https://github.com/bcoe/nyc#including-files), perhaps "specific files that should be covered".

The readme then shows glob patterns in an example without explicitly mentioning pattern support. So maybe add that too.

rpominov replied:
In readme the description is based on `exclude`, should we do the similar here after we add `exclude` too?

novemberborn replied:
Not sure. The include section in the readme follows the exclude section. I suspect users may view CLI usage information somewhat differently.

rpominov commented:
Also not 100% I did everything right, but I tested it in my project and it works fine.

novemberborn commented:
If we add `include` we should also add `exclude`.

rpominov commented on bin/nyc.js:
```diff
@@ -46,6 +46,11 @@ var yargs = require('yargs/yargs')(process.argv.slice(2))
   })
+  .option('include', {
+    alias: 'in',
+    default: [],
```
Also from readme:

> Note: include defaults to ['**']

But this default value probably defined somewhere else in the code. Should we remove it there, and specify `['**']` here. Although we need default value when it's loaded from package.json too, so maybe only specify here and not remove in other place?

novemberborn replied:
The code arrives at that `**` default implicitly. I think we'd need a guard [here](https://github.com/bcoe/nyc/blob/deee980b965cbe70b2da2c70b28ce9fc0b4a65d5/index.js#L40) for empty `config.include` arrays.

rpominov commented on bin/nyc.js:
```diff
@@ -46,6 +46,16 @@ var yargs = require('yargs/yargs')(process.argv.slice(2))
+  })
+  .option('include', {
+    alias: 'in',
+    default: ['**'],
```
I've set `['**']` so the help would like this:

```
--include, --in   a list of specific files that should be covered, glob
                  patterns are supported                     [default: ["**"]]
```

novemberborn replied:
`include` is used in [this line](https://github.com/bcoe/nyc/blob/3403ca17991fded623a05fda3e3398375320e1d9/index.js#L191):

```js
return (!this.include || micromatch.any(relFile, this.include)) && !micromatch.any(relFile, this.exclude)
```

I think with this change, when `nyc` is used through the CLI, it'll always call `micromatch.any`. Without a default it would just short-circuit the logic. We should definitely have the default in the documentation but maybe we should also [set in the constructor](https://github.com/bcoe/nyc/blob/3403ca17991fded623a05fda3e3398375320e1d9/index.js#L39:L42) so we can remove that logic, *or* set `this.include` to `false` if it's just `**`, so we can avoid the unnecessary match.

The difference between the CLI codepath and the embedding code path bothers me a little :smile:

rpominov commented:
I haven't check yet that the latest changes work. Will check in my project a bit later.

rpominov commented:
Checked:

- [x] `--all` and `include` not set: all files included
- [x] `--all --include **/src/**`: only src included
- [x] `--all` and `include: "**/src/**"` in package.json: only src included 
- [x] `--all` and `exclude: "**/src/**"` in package.json: src excluded
- [ ] `--all --exclude **/src/**`: src excluded

Looking into the last case.

rpominov commented:
The last commit fixes last case.

bcoe commented:
@rpominov I really like this work :+1: 

would it be possible to add add a test; no need to add one for the CLI (since we don't have any tests for this yet, although it would be great to have some eventually). But it would be nice to have tests for the new logic that appends a list of excludes provided to the default excludes, and logic for the behavior around an include that's an empty list.

Also, mind rebasing to resolve conflicts :+1:

rpominov commented:
@bcoe Ok, no problem. I think I'll be able to do this at the end of the week.

rpominov commented:
I've rebased and added the tests, hope they're ok.

Also I wasn't able to make all current tests pass on my machine (in master, without changes of this PR). In particular [this](https://github.com/bcoe/nyc/blob/3403ca17991fded623a05fda3e3398375320e1d9/test/src/nyc-test.js#L266) assertion fails, and [this](https://github.com/bcoe/nyc/blob/3403ca17991fded623a05fda3e3398375320e1d9/test/src/nyc-bin.js#L130) code fails with "file already exists" error. Probably this is just something wrong with my setup...

Also, man, running those tests splitted to individual files takes forever :hushed:

novemberborn commented:
LGTMO!

rpominov commented on bin/nyc.js:
```diff
@@ -109,9 +119,12 @@ if (argv._[0] === 'report') {
   var nyc = (new NYC({
-    require: argv.require
+    require: argv.require,
+    include: argv.include,
```
@novemberborn Maybe we could check for `**` here before passing to `new NYC`?

novemberborn replied:
Yea, or we change the `NYC` implementation. Let's see what @bcoe suggests.

bcoe commented:
@novemberborn @rpominov I made a couple small tweaks, and added tests. see:

https://github.com/rpominov/nyc/pull/4

rpominov commented:
@bcoe Thanks, merged. The commits log of this PR gets a little messy, should I squash the commits?

bcoe commented:
@rpominov no worries 👍 I've started using the squash button when I merge commits, along with this slick new library:

https://github.com/conventional-changelog/standard-version

rpominov commented:
Ah, right, forgot that github now can do squash. And standard-version looks very cool :+1:

bcoe commented:
@rpominov this is now released in `nyc@6.4.0`.

rpominov commented:
@bcoe Great. Thanks!
